### PR TITLE
Move loss calculation to head

### PR DESCRIPTION
**Pull request by AyushExel**
### Summary
🚀🛠️📝

This pull request adds a new feature to customize the loss function for detection models and refactors the code to use the new loss classes defined in `ultralytics/yolo/utils/loss.py`. It updates the documentation, simplifies the trainer scripts, and removes obsolete code.

> _The loss functions needed a change_
> _To make them more flexible and strange_
> _So they moved them to `loss.py`_
> _And made subclasses to apply_
> _To each task with a different range_

### Walkthrough
*  Add new loss function classes for different tasks that inherit from DetectionModel ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-7376f8301a51997a10901e6d8d5ef6a497f074f7c7d7209e003282245c4c6e2dR80-R368))
*  Import the new loss function classes in ultralytics/nn/tasks.py ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4R16))
*  Remove the criterion method from the BaseTrainer class and use the init_criterion method of the model instead ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-1dc4dec7a1716e080109cc732dc44c6e843b30bcdd324e0141e679ebf718c2b1L499-L504))
*  Add the init_criterion method to the ClassificationModel class and return an instance of the v8ClassificationLoss class ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-e2650ac65b6525f6d9c0594a226066d396afbcba637c054d29abe04e62f410c4L346-R351))
*  Remove the Loss, SegLoss, and criterion classes from the detect, segment, and pose modules and use the v8DetectionLoss, v8SegmentationLoss, and v8PoseLoss classes instead ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-c55f9a098f49780316e7ec36ccd2ae059a0dbf9129072e2cb9ce20bbbd5de940L138-L233), [link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-69b3b40601c35cc60cecb5a813fe62a36840599d895ebd439501a87a91f85581L62-L156), [link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-990260aed22cad0e3a42309be3fee54d7dabc589b0bd5558d1e47f0f32adcc70L48-L53))
*  Add a bullet point to the docs/usage/engine.md to document the customization of the loss function ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-cc3f027dd6bed4eb36cc55a8749053fc0003fe91b987000894baa43ec1c9a5adL51-R51))
*  Modify the example code in docs/usage/engine.md to show how to create a custom model class with the init_criterion method ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-cc3f027dd6bed4eb36cc55a8749053fc0003fe91b987000894baa43ec1c9a5adL57-R61))
*  Remove some unnecessary import statements and empty code blocks from various modules ([link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-c55f9a098f49780316e7ec36ccd2ae059a0dbf9129072e2cb9ce20bbbd5de940L5-L6), [link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-69b3b40601c35cc60cecb5a813fe62a36840599d895ebd439501a87a91f85581L4-L6), [link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-cc3f027dd6bed4eb36cc55a8749053fc0003fe91b987000894baa43ec1c9a5adL69), [link](https://github.com/ultralytics/ultralytics/pull/2874/files?diff=unified&w=0#diff-734f7c7d5e997138ddb17cffc50e37008364bb3fdba26403ee9d145ce2764f69L44))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces a major refactor of the Ultralytics neural network tasks, improving the forward pass and loss computation abstractions.

### 📊 Key Changes
- Added new loss classes to encapsulate various loss computations (`v8ClassificationLoss`, `v8DetectionLoss`, `v8PoseLoss`, `v8SegmentationLoss`).
- Refactored `BaseModel`, `DetectionModel`, `SegmentationModel`, `PoseModel`, `ClassificationModel`, and `RTDETRDetectionModel` forward methods for better abstraction.
- Removed `criterion` method from training scripts (`train.py`) and some old loss computation code.
- Improved handling of tensors for device compatibility in various utility functions.
- Made several forward pass and loss computation processes more concise and modular.

### 🎯 Purpose & Impact
- 🧠 Simplifies codebase by consolidating loss calculations into dedicated classes.
- 🚀 Enhances model extensibility for a variety of tasks (detection, classification, segmentation, pose estimation).
- 🛠 Facilitates easier debugging and maintenance of the codebase.
- 🔍 Improves forward pass methods, allowing better abstraction and clarity on prediction vs. loss calculation, which is beneficial for developers and users alike.

### CODE REVIEW

**Comment by AyushExel** [comment]
Okay there are some bug introduced in this branch when trying to rebase it against main. I'll need to debug it thoroughly. Will try tomorrow

**Comment by Laughing-q** [comment]
@AyushExel @glenn-jocher ok I've updated this PR and currently everything works fine I think.
- Added `BaseModel.predict()` and `BaseModel.loss()`, these functions can be called independently when it's predict mode or training mode.
- Modified `BaseModel.forward()` to call `.predict()` when the input is a tensor(predict mode) and call `.loss()` when the input is a dict(training mode).

Also I modified the loss and accuracy on both single-gpu training and ddp training with the following command:
```bash
yolo detect train model=yolov8n.pt data=coco128.yaml imgsz=640 epochs=10 close_mosaic=2   # for single-gpu
yolo detect train model=yolov8n.pt data=coco128.yaml imgsz=640 epochs=10 close_mosaic=2 device='0,1'   # for ddp
```
Here are the resutls for both, `new` means training with current branch and `base` means training with main branch.
The curves(losses and mAPs) are completely identical and coincident.
results
results-

**Comment by AyushExel** [comment]
@Laughing-q okay perfect! 
Let'a wait for Glenn to take a look and then merge this to main which will enable us to continue debugging rtdetr trainer

**Comment by glenn-jocher** [comment]
@Laughing-q this looks good, the only issue might be that I think _forward_once() is used with the `augment=True` flag for test-time augmentation in predict and val modes. This runs the image as 3 different sizes and combines the (de-augmented) outputs before NMS.

I'll check to see what's going on there.

**Comment by Laughing-q** [comment]
@glenn-jocher oh I didn't notice `forward_augment` is using `forward_once`. I've fixed it and tested predict mode which looks good.

**Comment by glenn-jocher** [comment]
@Laughing-q oh! Ok we should pass it during val then. Here's the YOLOv5 example from val.py:

Screenshot 2023-05-30 at 16 35 10

**Comment by glenn-jocher** [comment]
If we fix the pre-commit CI errors and add val augment then we are all good! I think val augment may only work for detect models.

**Comment by Laughing-q** [comment]
@glenn-jocher it seems currently classify does not support TTA. I'm heading to bed now, can you take a look? Thanks!
or maybe we just remove `augment=self.args.augment` in validator.py for now.

**Comment by glenn-jocher** [comment]
@Laughing-q PR merged!
